File: 0060-permutation-sequence/0060-permutation-sequence.js
/**
 * @param {number} n
 * @param {number} k
 * @return {string}
 */
var getPermutation = function(n, k) {
  const fact = n => {
    if (n === 0) return 1;
    return n * fact(n - 1);
  };
  
  const dfs = (q, s, ans=[]) => {
    if (s.length === 0) return ans;
    
    const subSize = fact(s.length - 1);
    const subIdx = Math.min(q / subSize);
    const ext = s.splice(subIdx, 1);
    return dfs(q % subSize, s, [...ans, ...ext]);
  }
  
  // const seq = _.range(1, n + 1);   * using lodash
  const seq = Array.from({ length: n }, (v, i) => (i + 1).toString());
  
  return dfs(k - 1, seq).join("");
};

File: 0060-permutation-sequence/0060-permutation-sequence.rb
# @param {Integer} n
# @param {Integer} k
# @return {String}
def fact(n)
  return 1 if n == 0
  return n*fact(n-1)
end

def dfs(q, s, ans='')
  return ans if s.length == 0
  
  sub_size = fact(s.length-1)
  sub_idx = q/sub_size
  ext = s.delete_at(sub_idx)
  dfs(q%sub_size, s, ans+ext)
end

def get_permutation(n, k)
  seq = (1..n).to_a.map{ |d| d.to_s }
  dfs(k-1, seq)
end